### ISSUE TITLE
Wavefront type-promotion in the CreateWavefront class

### ISSUE THREAD

**Issue by LouisDesdoigts** [LouisDesdoigts/dLux#60]
Currently the Opticalsystem instantiates a PhysicalWavefront without an option to change this behavior. Since we now have new wavefront types in the pipeline we want to be able to handle type promotion in the CreateWavefront class. I believe @Jordan-Dennis did some initial investigation.'

Other wavefront types will not be natively supported until this is completed.

**Comment by Jordan-Dennis** [comment]
So the thing I originally had in mind, did not work. `typing` provides generics, which are common in statically typed languages. For example in `java` there is the `List` container. But to compile a list, the compiler needs to know the data-type that it stores, however, the developers clearly weren't going to write a new implementation for every data-type; plus what happens for a `List` of user defined types. The answer is generics which can be used to indicate an unknown type at compile time. Hence you can have `List<Integer>` or `List<Fish>`. I thought we could use this and have `CreateWavefront[PhysicalWavefront]` or `CreateWavefront[AngularWavefront]` which is exactly what we need. Unfortunately `typing.GenericAlias` seems to only be for type annotations and does not actually provide functionality.

**Comment by Jordan-Dennis** [comment]
`typing` also provides `cast` but this just changes the `RunTime` type that the type-checker sees and will not change anything in memory. I think that the best way to go about this would probably be one of four things; a) use `Generic` but with our own logic behind the scenes or b) implement multiple `CreateWavefront` style layers, c) implement `CastWavefront(T, V)` as a layer to cast a wavefront from `T` to `V` or d) implement a `cast(T)` method on each wavefront, that would contain much less logic than c). 

All of these would be generic methods using `Generic[T]` for the type signatures. a) would look something like:
```py
class CreateWavefront(eqx.Module, Generic[T]):
    # __init__ and parameters
    def __call__(self, parameters : dict) -> dict:
        if T == type(GaussianWavefront):
            # Do x thing
        elif T == type(AngualrWavefront):
            # Do y thing
        else: # default
            # PhysicalWavefront
```
I discussed b) in #59 so I won't elaborate anymore here. c) is reasonably straight forward and uses a lot of the machinery I described above, only it will need an `if`, `elif`, `else` nested twice for the 9 possible casts which could be quite complex, especially if we want to use `jax` logic functions. d) would look something along the lines of a `Generic` method:
```py
def cast(self, wavefront : T) -> T:
    jax.lax.cond(T == Type(ReferenceWavefront), 
        lambda wavefront : _cast_to_reference_wavefront(wavefront),
        lambda wavefront : _cast_to_other_option(wavefront),
        self)
```
Which is `jax` logic capable, hence why I included it above in this form. Let me know what option seems best or if you have a better one.

**Comment by LouisDesdoigts** [comment]
Yeah I actually think we might have an even simpler solution will make out life easier. Rather than construct the WaveFront object in the `OpticalSystem.propagate_mono()` and putting it in the params_dict, we just pass the wavelength and offset values through the params_dict and instantiate the wavefront in the CreateWavefront class using basic python logic. Then we can just pass a simple `wf_type=str` kwarg into the CreateWavefront class and were done.

**Comment by Jordan-Dennis** [comment]
Yeah that sounds good. Nice and simple. I think d) might still have merit because passing a planar wavefront, (i.e. `PhysicalWavefront`) through a `GaussianMirror` we might want to cast it to `GaussianWavefront`. I guess the other option is just to use `GaussianWavefront` instead of the `PhysicalWavefront` through the entire system as this will work. When planar the `phase_radius = np.inf` would be my guess.

**Comment by LouisDesdoigts** [comment]
Yeah so I don't think thats an issue - The propagator/wavefront pair needs to match, we can't use a PhysicalWavefront on a GaussianPropagator because they don't track the right values. It should be trivial to create layers to simply load the data from one wavefront type and initialise a new one with the same properties in order to change the propagation types along the optical path. There's obviously some nuance in edge cases there, but I think this is one of these corners where auto-diff mandates the garbage-in garbage-out philosophy. We can always add basic python checks/warning once we know some of the edge/corner cases.

**Comment by LouisDesdoigts** [comment]
Temporary solution: Pass wavelength and offset as parameters inside params_dict and initialise the wavefront entirely inside the `CreateWavefront` layer. Will do this soon so we can have full wavefront functionality, and then we can use proper type promotion later.

**Comment by benjaminpope** [comment]
PS I think we want to say Cartesian instead of Physical...

**Comment by LouisDesdoigts** [comment]
Fixed with commit #1dc3957
